docs/design/three_datahall.md

```diff
@@ -0,0 +1,204 @@
+The current requirement for `three_data_hall` and `three_datacenter` is to have 3 AZs and at least two different failure zones per `data_hall`/`datacenter`.
+
+## General Design Goals
+
+The goal of this design is to support the redundancy modes `three_data_hall` and `three_datacenter` in different deployment models:
+
+* Single Kubernetes cluster across multiple AZs.
+* One Kubernetes cluster per AZ.
+
+The implementation should be flexible enough to give the user a choice of the used AZ.
+
+## Current Implementation
+
+The current version of the operator doesn't support this redundancy mode.
+The locality is currently a mix of some constants, a configurable zone and two settings at the cluster level [locality configuration](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/master/controllers/cluster_controller.go#L631-L642).
+
+## Proposed Design
+
+The design is split into two main parts of the deployment: configure the locality and how to deploy the Pods across the cluster.
+
+### Configure locality
+
+The `three_data_hall` deployment requires the `locality_data_hall` locality to be set and the `three_datacenter` requires the `locality_dcid` to be set.
+There must be at least and at most 3 different values for these localities.
+To full fill the requirements there must be at least two different `locality_zone`s per `data_hall`/`datacenter`.
+The `locality_zone` can be configured with [FoundationDBClusterFaultDomain](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/master/docs/cluster_spec.md#foundationdbclusterfaultdomain).
+In addition to that the user can configure additional variables that the sidecar will use for substitution in the [FoundationDBClusterSpec](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/master/docs/cluster_spec.md#foundationdbclusterspec) with the `sidecarVariables` list.
+The current implementation would allow to set `dataHall: $AZ` and in `sidecarVariables` we would list `AZ` to define the `locality_data_hall` based on an environment variable that will be replaced by the sidecar with the actual value.
+Instead of having these different mechanisms spread across different settings I propose to have a new `localities` setting in the `FoundationDBClusterSpec`.
+The key will always be prefixed with `locality_`:
+
+```yaml
+localities:
+- key: "data_hall"
+  value: ""
+  valueFromEnv: ""
+  valueFromNode: "topology.kubernetes.io/zone"
+  topologyKey: ""
+- key: "zone"
+  value: ""
+  valueFromEnv: "MyFancyZone"
+  valueFromNode: ""
+  topologyKey: "kubernetes.io/hostname"
+```
+
+If multiple fields of the `value`, `valueFromEnv` or `valueFromNode` are set the following order will be used:
+
+1. `value`
+1. `valueFromEnv`
+1. `valueFromNode`
+
+The `topologyKey` will be used for `topologySpreadConstraints` and for `PodAntiAffinity` and only has to be set if `valueFromNode` is empty, otherwise it will default to that value.
```

Comment:
Could we combine the `topologyKey` and `valueFromNode` fields, or do we think there is a use case for them to be different? I see the zone example where we have `valueFromNode` blank, but since `valueFromNode` is the lowest priority option in the different `value...` fields, I think it would have the same behavior whether it was set or not in that case.

Reply:
Conclusion from discussion: combine them.